File: modules/rust_workloads/src/pointer_chase.rs
//! Pointer chasing over a single-cycle permutation versus a contiguous sum of
//! the same buffer: dependent random loads against streaming loads.

use rand::rngs::StdRng;
use rand::{Rng, SeedableRng};

/// Single-cycle permutation of 0..n (Sattolo's algorithm): following
/// `i = perm[i]` from any start visits every slot exactly once.
pub fn build_cycle(n: usize, seed: u64) -> Vec<usize> {
    let mut rng = StdRng::seed_from_u64(seed);
    let mut perm: Vec<usize> = (0..n).collect();
    for i in (1..n).rev() {
        let j = rng.random_range(0..i);
        perm.swap(i, j);
    }
    perm
}

/// Walk the cycle once from slot 0, summing the visited indices.
pub fn chase_sum(perm: &[usize]) -> usize {
    let mut idx = 0;
    let mut acc = 0;
    for _ in 0..perm.len() {
        acc += idx;
        idx = perm[idx];
    }
    acc
}

/// Contiguous sum of the same buffer — the streaming counterpart.
pub fn array_sum(perm: &[usize]) -> usize {
    perm.iter().sum()
}

File: modules/java_workloads/src/PointerChase.java
package monorepo.workloads;

import java.util.SplittableRandom;

/**
 * Pointer chasing over a single-cycle permutation versus a contiguous
 * sum of the same buffer: dependent random loads against streaming loads.
 */
@SuppressWarnings("PMD.ShortVariable")
public final class PointerChase {
    private PointerChase() {
    }

    /**
     * Single-cycle permutation of 0..n (Sattolo's algorithm): following
     * i = perm[i] from any start visits every slot exactly once.
     */
    public static int[] buildCycle(int n, long seed) {
        SplittableRandom rng = new SplittableRandom(seed);
        int[] perm = new int[n];
        for (int i = 0; i < n; i++) {
            perm[i] = i;
        }
        for (int i = n - 1; i >= 1; i--) {
            int j = rng.nextInt(i);
            int tmp = perm[i];
            perm[i] = perm[j];
            perm[j] = tmp;
        }
        return perm;
    }

    /**
     * Walks the cycle once from slot 0, summing the visited indices.
     */
    public static long chaseSum(int[] perm) {
        int idx = 0;
        long acc = 0;
        for (int unused : perm) {
            acc += idx;
            idx = perm[idx];
        }
        return acc;
    }

    /**
     * The streaming counterpart: a contiguous sum of the buffer.
     */
    public static long arraySum(int[] perm) {
        long acc = 0;
        for (int v : perm) {
            acc += v;
        }
        return acc;
    }
}
